apis/memorystore/v1alpha1/instance_types.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	refs "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/apis/k8s/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var MemorystoreInstanceGVK = GroupVersion.WithKind("MemorystoreInstance")
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// MemorystoreInstanceSpec defines the desired state of MemorystoreInstance
+// +kcc:proto=google.cloud.memorystore.v1beta.Instance
+type MemorystoreInstanceSpec struct {
+	// Optional. Immutable. The MemorystoreInstance name. If not given, the metadata.name will be used.
+	ResourceID *string `json:"resourceID,omitempty"`
+
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Parent field is immutable"
+	// Immutable.
+	// +required
+	Parent `json:",inline"`
+
+	// Optional. Number of replica nodes per shard. If omitted the default is 0
+	//  replicas.
+	ReplicaCount *int32 `json:"replicaCount,omitempty"`
+
+	// Optional. Immutable. Authorization mode of the instance.
+	AuthorizationMode *string `json:"authorizationMode,omitempty"`
+
+	// Optional. Immutable. In-transit encryption mode of the instance.
+	TransitEncryptionMode *string `json:"transitEncryptionMode,omitempty"`
```

Comment:
isn't this field required? https://cloud.google.com/memorystore/docs/valkey/reference/rest/v1beta/projects.locations.instances#Instance

Reply:
Both Authorization mode and TransitEncryptionMode are optional

Reply:
As per discussion, the shardcount field is actually optional. The public doc needs to be updated